ArmPkg/Library/ArmFfaLib/ArmFfaRxTxMap.c

```diff
@@ -0,0 +1,239 @@
+
+**/
+#include <Uefi.h>
+#include <Pi/PiMultiPhase.h>
+
+#include <Library/ArmLib.h>
+#include <Library/ArmFfaLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/PcdLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+
+#include <IndustryStandard/ArmFfaSvc.h>
+
+#include "ArmFfaCommon.h"
+#include "ArmFfaRxTxMap.h"
+
+/**
+  Get mapped Rx/Tx buffers.
+
+  @param [out]   TxBuffer         Address of TxBuffer
+  @param [out]   TxBufferSize     Size of TxBuffer
+  @param [out]   RxBuffer         Address of RxBuffer
+  @param [out]   RxBufferSize     Size of RxBuffer
+
+  @retval EFI_SUCCESS
+  @retval Others             Error.
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibGetRxTxBuffers (
+  OUT VOID    **TxBuffer,
+  OUT UINT64  *TxBufferSize,
+  OUT VOID    **RxBuffer,
+  OUT UINT64  *RxBufferSize
+  )
+{
+  if ((PcdGet64 (PcdFfaTxBuffer) == 0x00) || (PcdGet64 (PcdFfaRxBuffer) == 0x00)) {
+    return EFI_NOT_READY;
+  }
+
+  if (TxBuffer != NULL) {
+    *TxBuffer = (VOID *)(UINTN)PcdGet64 (PcdFfaTxBuffer);
+  }
+
+  if (TxBufferSize != NULL) {
+    *TxBufferSize = PcdGet64 (PcdFfaTxRxPageCount) * EFI_PAGE_SIZE;
+  }
+
+  if (RxBuffer != NULL) {
+    *RxBuffer = (VOID *)(UINTN)PcdGet64 (PcdFfaRxBuffer);
+  }
+
+  if (RxBufferSize != NULL) {
+    *RxBufferSize = PcdGet64 (PcdFfaTxRxPageCount) * EFI_PAGE_SIZE;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Mapping Rx/Tx buffers.
+  This function is only called in ArmFfaLibConstructor because
+  Rx/Tx buffer is registered only once per partition.
+
+  @retval EFI_SUCCESS
+  @retval EFI_ALREADY_STARTED     Rx/Tx buffer already mapped.
+  @retval EFI_OUT_OF_RESOURCE     Out of memory
+  @retval EFI_INVALID_PARAMETER   Invalid alignment of Rx/Tx buffer
+  @retval Others                  Error
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibRxTxMap (
+  IN VOID
+  )
+{
+  EFI_STATUS    Status;
+  ARM_FFA_ARGS  FfaArgs;
+  UINTN         MinSizeAndAlign;
+  UINTN         Property2;
+  VOID          *Buffers;
+  VOID          *TxBuffer;
+  VOID          *RxBuffer;
+  UINT64        BufferSize;
+
+  TxBuffer   = (VOID *)(UINTN)PcdGet64 (PcdFfaTxBuffer);
+  RxBuffer   = (VOID *)(UINTN)PcdGet64 (PcdFfaRxBuffer);
+  BufferSize = PcdGet64 (PcdFfaTxRxPageCount) * EFI_PAGE_SIZE;
+
+  /*
+   * If someone already mapped Rx/Tx Buffers, return EFI_ALREADY_STARTED.
+   * return EFI_ALREADY_STARTED.
+   */
+  if ((TxBuffer != NULL) && (RxBuffer != NULL)) {
+    return EFI_ALREADY_STARTED;
+  }
+
+  Status = ArmFfaLibFeatures (ARM_FID_FFA_RXTX_MAP, 0x00, &MinSizeAndAlign, &Property2);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to get RX/TX buffer property... Status: %r\n",
+      __func__,
+      Status
+      ));
+    return Status;
+  }
+
+  ZeroMem (&FfaArgs, sizeof (ARM_FFA_ARGS));
+
+  switch (MinSizeAndAlign) {

```

Comment:
Spec definition is that "Bit[1:0]: Minimum buffer size and alignment boundary". this will fail a lot of use cases.

Reply:
> Spec definition is that "Bit[1:0]: Minimum buffer size and alignment boundary". this will fail a lot of use cases.

That's why I think user should call FFA_FEATURES before calling this API. 
and it's " Minimum buffer size and alignment boundary" for RX/TX buffer. so I don't think it fails in usecase?

Reply:
Sorry, what I meant was we should bit and only switch case with those lower 2 bits. When testing it, I received 0x10000 in this query, which should be a legitimate return, but failed here.

Reply:
Oh. Thanks a lot. That's horrible my mistake :)